File: src/Shenora.React/src/appLifecycle.ts
/**
 * When the APP went away and came back — and how long it was gone, which is the part this page cannot
 * measure itself. Mirrors `Shenora.Modules.Platform.AppLifecycle`, pinned by `WireMirrorTests`.
 *
 * 🔴 **For "am I on screen", use `document.visibilitychange`.** It fires on both mobile shells and is
 * the web platform's own answer; this would only arrive later and over IPC. What it adds is the two
 * things a hidden document genuinely cannot know:
 *
 * 1. **How long it was away, on a clock that was not throttled.** A backgrounded page's timers are
 *    throttled and its process may be frozen, so a `Date.now()` delta across the gap is unreliable —
 *    and the duration is what the decision actually turns on. Three seconds in the notification shade
 *    needs no reconnect; forty minutes means the socket is dead.
 * 2. **That this was the user leaving the APP**, rather than anything else that can hide a document.
 *
 * ⚠ It reports; it does not act. Reconnecting, re-probing or refetching is yours — the host has no way
 * to know what this page was holding.
 */
import { useEffect, useRef } from 'react';
import { eventBus as defaultEventBus, type ShenoraEventBus } from './eventBus.js';

/** The module these events are published under. */
export const LIFECYCLE_MODULE = 'SHENORA.LIFECYCLE';

/** The app left the foreground. No payload. */
export const LIFECYCLE_STOPPED = 'STOPPED';

/** The app came back, carrying an {@link AppLifecycleReport}. */
export const LIFECYCLE_RESUMED = 'RESUMED';

/** What a {@link LIFECYCLE_RESUMED} event carries. */
export interface AppLifecycleReport {
  /**
   * How long the app was away, or `null` when there was no preceding stop to measure from — a first
   * launch, or a shell that reported only the resume.
   *
   * ⚠ **`null` is not `0`.** Zero would be a measurement; this is the absence of one. Treating them
   * alike makes the first resume after launch skip the reconnect every other resume performs.
   */
  backgroundMilliseconds: number | null;
}

/** What {@link useAppLifecycle} watches for. Every handler is optional. */
export interface AppLifecycleHandlers {
  /** The app left the foreground. */
  onStopped?: () => void;
  /**
   * The app came back. `backgroundMilliseconds` is null when there was nothing to measure.
   *
   * ⚠ Branch on a THRESHOLD rather than on the event: reconnecting on every resume makes switching
   * apps expensive, and doing nothing makes a long absence look like a hung page.
   */
  onResumed?: (report: AppLifecycleReport) => void;
}

/**
 * Run something when the app leaves the foreground and when it comes back.
 *
 * ```tsx
 * useAppLifecycle({
 *   onResumed: ({ backgroundMilliseconds }) => {
 *     // A short trip to the notification shade costs nothing; a long one means the socket is gone.
 *     if (backgroundMilliseconds === null || backgroundMilliseconds > 30_000) reconnect();
 *   },
 * });
 * ```
 *
 * ⚠ On a shell that reports no lifecycle — the desktop, or a mobile app that did not compose
 * `MobileAppLifecycle` — neither handler ever runs. There is no capability to branch on, because there
 * is nothing to render differently: an app that also wants a visibility signal should use
 * `document.visibilitychange`, which works everywhere including a plain browser tab.
 */
export function useAppLifecycle(
  handlers: AppLifecycleHandlers,
  options: { eventBus?: ShenoraEventBus } = {},
): void {
  const eventBus = options.eventBus ?? defaultEventBus;
  // Read through a ref so a handler rebuilt every render — the normal case for inline arrows — does not
  // tear down and rebuild the subscription, losing any transition that lands in the gap.
  const latest = useRef(handlers);
  latest.current = handlers;

  useEffect(() => {
    const subscriptions = [
      eventBus.subscribe(LIFECYCLE_MODULE, LIFECYCLE_STOPPED, () => {
        latest.current.onStopped?.();
      }),
      eventBus.subscribe<AppLifecycleReport>(LIFECYCLE_MODULE, LIFECYCLE_RESUMED, (message) => {
        // ⚠ Absent rather than 0 when the host sent nothing usable — the same distinction the payload
        // documents, preserved here so a missing field cannot read as "away for no time".
        const reported = message.payload?.backgroundMilliseconds;
        latest.current.onResumed?.({
          backgroundMilliseconds: typeof reported === 'number' ? reported : null,
        });
      }),
    ];
    return () => subscriptions.forEach((off) => off());
  }, [eventBus]);
}

File: src/Shenora/Modules/Platform/AppLifecycle.cs
using System.Diagnostics;
using Microsoft.Extensions.Logging;
using Shenora.Core.Events;

namespace Shenora.Modules.Platform;

/// <summary>
/// Tells the page when the APP went away and came back, and — the part it cannot measure itself — how
/// long it was gone.
/// <para>
/// 🔴 <b>Use <c>document.visibilitychange</c> for "am I on screen".</b> It fires on both shells and is
/// the web platform's own answer; a kit event duplicating it would be worse, because it arrives later
/// and over IPC. This exists for the two things a hidden page genuinely cannot know:
/// </para>
/// <list type="number">
/// <item>
/// <b>HOW LONG it was away, measured by a clock that was not throttled.</b> A backgrounded page's timers
/// are throttled and its process may be frozen outright, so <c>Date.now()</c> deltas taken across the gap
/// are unreliable — and the decision an app actually makes on resume turns on the duration. Three seconds
/// in the notification shade needs no reconnect; forty minutes means the socket is dead and whatever it
/// was paired with may have come or gone.
/// </item>
/// <item>
/// <b>That this was an ACTIVITY transition</b> — MAUI's <c>Window.Stopped</c>/<c>Resumed</c>, i.e. the
/// user leaving the app — rather than anything else that can hide a document.
/// </item>
/// </list>
/// <para>
/// ⚠ <b>It reports; it does not act.</b> What to do about a long absence is the app's — reconnect,
/// re-probe, refetch, or nothing. The kit has no way to know what the page was holding.
/// </para>
/// </summary>
/// <remarks>
/// Portable: the shell supplies the two transitions and this owns the timing, so the arithmetic that
/// decides an app's reconnect is testable with no device. <c>MobileAppLifecycle</c> is the platform half.
/// </remarks>
public sealed class AppLifecycle
{
    /// <summary>The module these events are published under.</summary>
    public const string Module = "SHENORA.LIFECYCLE";

    /// <summary>Event: the app left the foreground. No payload.</summary>
    public const string StoppedType = "STOPPED";

    /// <summary>Event: the app came back, carrying an <see cref="AppLifecycleReport"/>.</summary>
    public const string ResumedType = "RESUMED";

    private readonly IEventBus _events;
    private readonly ILogger? _log;
    private readonly Lock _gate = new();

    // Monotonic, so a clock change mid-background cannot produce a negative or absurd duration — which a
    // page WOULD act on, since the whole payload is a number it branches against a threshold.
    private long? _stoppedAt;

    /// <param name="events">Where the transitions are published. The pump forwards them to the page.</param>
    /// <param name="log">Optional diagnostics.</param>
    public AppLifecycle(IEventBus events, ILogger? log = null)
    {
        _events = events ?? throw new ArgumentNullException(nameof(events));
        _log = log;
    }

    /// <summary>
    /// The app left the foreground. Call from the shell's stop signal — on MAUI that is
    /// <c>Window.Stopped</c>.
    /// <para>
    /// ⚠ <b><c>Stopped</c>, not <c>Deactivated</c>.</b> The latter also fires for a dialog or the
    /// notification shade, which would report the app as backgrounded while it is still on screen — and
    /// a page reconnecting a socket every time a permission prompt appears is worse than one that never
    /// reconnects at all.
    /// </para>
    /// </summary>
    public void ReportStopped()
    {
        lock (_gate)
        {
            // Idempotent: a second stop without an intervening resume keeps the FIRST timestamp, so the
            // duration still covers the whole absence rather than restarting mid-way.
            _stoppedAt ??= Stopwatch.GetTimestamp();
        }
        Log("lifecycle: the app left the foreground");
        _events.Emit(Module, StoppedType);
    }

    /// <summary>
    /// The app came back. Call from the shell's resume signal — on MAUI that is <c>Window.Resumed</c>.
    /// <para>
    /// ⚠ A resume with no preceding stop reports <see cref="AppLifecycleReport.BackgroundMilliseconds"/>
    /// as null rather than zero. Zero would be a measurement, and this is the absence of one — the first
    /// resume after launch is exactly that case, and a page treating it as "away for 0 ms" would skip the
    /// reconnect it does on every other resume.
    /// </para>
    /// </summary>
    public void ReportResumed()
    {
        double? awayMs;
        lock (_gate)
        {
            awayMs = _stoppedAt is { } since
                ? Stopwatch.GetElapsedTime(since).TotalMilliseconds
                : null;
            _stoppedAt = null;
        }

        Log(awayMs is { } ms
            ? $"lifecycle: the app returned after {ms:0}ms in the background"
            : "lifecycle: the app is in the foreground (no preceding stop — a first launch or a resume "
            + "the shell did not report)");
        _events.Emit(Module, ResumedType, new AppLifecycleReport(awayMs));
    }

    private void Log(string message) => AppCallback.Log(_log, () => $"[Shenora] {message}");
}

/// <summary>
/// What an <see cref="AppLifecycle.ResumedType"/> event carries.
/// </summary>
/// <param name="BackgroundMilliseconds">
/// How long the app was away, or null when there was no preceding stop to measure from — a first launch,
/// or a shell that reported only the resume. ⚠ Null is not zero: see
/// <see cref="AppLifecycle.ReportResumed"/>.
/// </param>
public sealed record AppLifecycleReport(double? BackgroundMilliseconds);
